flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java

```diff
@@ -142,4 +143,16 @@
      */
     String getSelectFromStatement(
             String tableName, String[] selectFields, String[] conditionFields);
+
+    /** Create catalog instance. */
+    default AbstractJdbcCatalog createCatalog(
+            String catalogName,
```

Comment:
We should move all of these parameters into a `Context` interface and pass that so we can easily add more information if we need to without breaking the API. This is similar to how table factories work, and basically the JDBC dialect infrastructure is just a clone of table factories anyway.

Reply:
u mean we should revise from here `https://github.com/apache/flink/blob/d753cf20826c66f20bfbd3772c58709e00d9bcb7/flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/catalog/factory/JdbcCatalogFactory.java#L72` ?

Reply:
Yeah, I think passing the original `CatalogFactory#Context` would be best as that gives implementations access to everything a normal catalog factory would have access to, like the configuration and classloader. We should also annotate `JdbcCatalogFactoryOptions` as `@PublicEvolving` then so implementations can use it to retrieve catalog options.

I'm now seeing that `JdbcCatalog` is already public and hard-wired to use username/password. That is unfortunate.

Reply:
ok，i prefer to create new issue for `Context`. of course, if u think it is better to update `Context` in the patch, i can modify it

Reply:
Since we are creating public API in this PR, we need to make sure the API is sound in this PR and not defer it to another task.

Reply:
but the JdbcDialect already has `@PublicEvolving`, https://github.com/apache/flink/blob/d753cf20826c66f20bfbd3772c58709e00d9bcb7/flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java#L35

Reply:
But you are adding a new public API method to this interface in this PR.

Reply:
> But you are adding a new public API method to this interface in this PR. So we need to make sure that the design of this method is good.

sorry, i dont understand how I should do it... Can you go into more detail?

Reply:
Basically, I think we should just pass the `context` argument from `JdbcCatalogFactory#createCatalog` directly to `JdbcDialect#createCatalog` instead of "destructuring" it into into a list of the individual options. Then any changes to `CatalogFactory#Context` would also automatically become available for dialect implementations.

I think we should also remove `username`, `pwd`, `baseUrl`, `defaultUrl`, `getUsername`, `getPassword`, and `getBaseUrl` from `AbstractJdbcCatalog`. We can either move the connection test logic of `#open` into `JdbcCatalogUtils` and call it from `PostgresCatalog`, or we instead add a `abstract Connection getConnection()` method to `AbstractJdbcCatalog` and use that in `#open`. This decouples the catalog implementation from the concrete authentication implementation.

Reply:
Maybe CC @bowenli86 and @JingsongLi to give feedback on this, too.

Reply:
> Basically, I think we should just pass the `context` argument from `JdbcCatalogFactory#createCatalog` directly to `JdbcDialect#createCatalog` instead of "destructuring" it into into a list of the individual options. Then any changes to `CatalogFactory#Context` would also automatically become available for dialect implementations.
> 
> I think we should also remove `username`, `pwd`, `baseUrl`, `defaultUrl`, `getUsername`, `getPassword`, and `getBaseUrl` from `AbstractJdbcCatalog`. Instead add a `abstract Connection getConnection()` method to `AbstractJdbcCatalog` and use that in `#open`. This decouples the catalog implementation from the concrete authentication implementation.
> 
> But I'm happy to hear other people's thoughts on this, too.

thx @Airblader  got it. but i prefer to create a new refactored issue

Reply:
I'd be OK with doing this in a separate ticket, but then that ticket needs to be done first before we can merge this. This creates new public APIs, and public APIs need to be correct when they are merged. Otherwise the other ticket will fall through the cracks and this gets released as-is. I'm not a committer, though, this is just my opinion.

Reply:
> Otherwise the other ticket will fall through the cracks and this gets released as-is. I'm not a committer, though, this is just my opinion.

i am sorry, i dont understand why can not we merge the PR first and then create a new ticket?

Reply:
Once a new public API is released, we cannot just change it anymore – because it is a breaking change for users, and we need to avoid breaking changes. This is different from internal implementations which we can change whenever we need to. In this case, the internal implementation doesn't make for a good public API, so we need to bring it into better shape. If we just open another issue, we cannot ensure that this issue would be resolved prior to the next release.

Reply:
The [current discussion](https://lists.apache.org/x/thread.html/ra438e89eeec2d4623a32822e21739c8f2229505522d73d1034e34198@%3Cdev.flink.apache.org%3E) on the dev ML really emphasizes that we simply need to put more focus and effort into designing APIs well and avoid breaking them.

Reply:
Yes, you are right. 
But I dont understand why you have to refactor the code on this PR. This PR is designed to support custom calalogs :-)

Reply:
Because AbstractJdbcCatalog is currently not a public API, but with your PR it becomes one. Therefore it must have a good design or else we'll be forced to break the API in the future, which is exactly what we need to avoid. I cannot think of more ways to say this. Why do you want to move this to a separate issue? It only creates a big risk of releasing a bad API because no one implements the follow-up issue. Things should have the proper quality when they hit master, not when a release is created.